### PR TITLE
Add template to custom types and complex props

### LINKED ISSUE #143: Add corresponding -Component and -Render props for nested options templates
### Issue Type
feature request

### Description
This relates to all nested options template props. E.g. `DataGrid`'s `Column.cellTemplate`.

#### Current Behavior
The `Column` component has only the `cellTemplate` prop

#### Expected Behavior
The `Column` component has `cellRender` and `cellComponent` as well

### Environment Details

- devextreme-react version: 18.1.4-alpha.8

### PR DESCRIPTION
**Pull request by lukyanovas**
resolves #143

### CODE REVIEW

File: src/core/template-helper.ts [hunk 1]
```diff
@@ -14,57 +17,86 @@ interface IDxTemplate {
...
-interface ITemplateInfo {
+interface ITemplateBaseDto {
     name: string;
     prop: string;
     isNested: boolean;
     isComponent: boolean;
+    propsGetter: (prop: string) => any;
+}
+
+interface ITemplateDto extends ITemplateBaseDto {
```
Comment (IlyaKhD) [inline src/core/template-helper.ts]: What about naming it `ITemplateDescriptor`? -`Dto` is from DB\ORM\etc world, personally I like `Descriptor` (or `Descr`) more

File: src/core/component.ts [hunk 2]
```diff
@@ -34,26 +31,25 @@ abstract class ComponentBase<P> extends React.PureComponent<P, IState> {
...
       templates: {}
     };
 
     this._optionsManager = new OptionsManager((name) => this.props[name]);
-    this._templateHelper = new TemplateHelper(this);
   }
 
   public componentWillUpdate(nextProps: P) {
-    const preparedProps = this._prepareProps(nextProps);
+    const preparedProps = this._prepareProps(nextProps, this._defaults, this._templateProps);
```
Comment (IlyaKhD) [inline src/core/component.ts]: As the `_prepareProps` is an instance method, is in necessary to pass `_defaults` and `_templateProps` as params?

File: src/core/component.ts [hunk 3]
```diff
@@ -105,115 +97,82 @@ abstract class ComponentBase<P> extends React.PureComponent<P, IState> {
 
   protected _createWidget(element?: Element) {
     element = element || this._element;
-    const nestedProps = this._optionsManager.getNestedOptionsObjects();
+    const nestedProps = this._optionsManager.getNestedOptionsObjects(this._updateState.bind(this));
```
Comment (IlyaKhD) [inline src/core/component.ts]: `this._updateState` was already bound to `this` already (`ctor`)

File: src/core/component.ts [hunk 4]
```diff
@@ -105,115 +97,82 @@ abstract class ComponentBase<P> extends React.PureComponent<P, IState> {
...
-        ...separatedProps,
-        options,
-        nestedTemplates
-    };
-  }
+    const templateOptions = getTemplateOptions({
+      templateProps,
+      options: separatedProps.templates,
+      nestedOptions: getNestedTemplates(rawProps),
+      stateUpdater: this._updateState.bind(this),
```
Comment (IlyaKhD) [inline src/core/component.ts]: `this._updateState.bind` was already bound in the `ctor`

File: src/core/component.ts [hunk 5]
```diff
@@ -105,115 +97,82 @@ abstract class ComponentBase<P> extends React.PureComponent<P, IState> {
...
 
     this._instance = new this._WidgetClass(element, options);
     this._optionsManager.setInstance(this._instance);
     this._instance.on("optionChanged", this._optionsManager.handleOptionChange);
   }
 
-  private _prepareProps(rawProps: Record<string, any>): IWidgetConfig {
-    const separatedProps = separateProps(rawProps, this._defaults, this._templateProps);
-    let options = separatedProps.options;
+  private _updateState(callback) {
```
Comment (IlyaKhD) [inline src/core/component.ts]: rename it to `_updateTemplatesState` ?

File: src/core/component.ts [hunk 6]
```diff
@@ -105,115 +97,82 @@ abstract class ComponentBase<P> extends React.PureComponent<P, IState> {
...
 
-  private _wrapEventHandler(optionValue: any, optionName: string): any {
-      if (optionName.substr(0, 2) === "on" && typeof optionValue === "function") {
-          return (...args: any[]) => {
-              if (!this._optionsManager.updatingProps) {
-                  optionValue(...args);
-              }
+function getNestedTemplates(rawProps: Record<string, any>): Record<string, any> {
+  let nestedTemplates: Record<string, any> = {};
+  if (rawProps.children) {
```
Comment (IlyaKhD) [inline src/core/component.ts]: Can the check be omitted as on the line 65?

File: src/core/template-helper.ts [hunk 7]
```diff
@@ -14,57 +17,86 @@ interface IDxTemplate {
...
+    templates: any;
+    templateStubs: any;
+} {
+    const templates: Record<string, IDxTemplate> = {};
+    const templateStubs: Record<string, any> = {};
+    const options = meta.options;
+    const stateUpdater = meta.stateUpdater;
+
+    meta.templateProps.forEach((m) => {
+        if (options[m.component] || options[m.render]) {
```
Comment (IlyaKhD) [inline src/core/template-helper.ts]: Invert condition and continue?

File: src/core/component.ts [hunk 8]
```diff
@@ -2,27 +2,24 @@ import * as events from "devextreme/events";
...
-import { ITemplateMeta, Template } from "./template";
+import { findProps as findTemplateProps, ITemplateMeta } from "./template";
 import { separateProps } from "./widget-config";
 
 import {
-  IDxTemplate,
-  ITemplateInfo,
-  IWrappedTemplateInfo,
-  TemplateHelper
+  getTemplateOptions,
```
[the remark this answers is not in the document]
Comment (lukyanovas) [reply src/core/component.ts]: I'm going to remove `findTemplateProps` function in next PR's

[the remark this answers is not in the document]
Comment (IlyaKhD) [reply src/core/component.ts]: Nevertheless if this is not going to be included into this PR, better rename it know, just for sure. Maybe this next PR occurs in a few weeks, who knows?

**Pull request merged**